index.js

```diff
@@ -104,12 +113,16 @@ function compression (options) {
       }
 
       if (!stream) {
-        return _end.call(this, chunk, encoding)
+        return endOnce.call(this, chunk, encoding)
       }
 
       // mark ended
       ended = true
 
+      if (onFinished.isFinished(this)) {
+        return endOnce.call(this)
+      }
```

Comment:
This addition takes care of situations where a route handler calls `res.end()` _after_ the response is considered "finished". This is often due to the client having moved on before the response is ready for them.